### ISSUE TITLE
[MPS] `out_channels` <= 2**16 guard for convolution is too broad

### ISSUE THREAD

**Issue by hvaara** [pytorch/pytorch#142515]
### 🐛 Describe the bug

In https://github.com/sdatkinson/neural-amp-modeler/issues/505, @sdatkinson discovered that the guard against `out_channels` > 2**16 in convolution

https://github.com/pytorch/pytorch/blob/117b6c3e2c477793e3b5f964ec01bd0718313aae/aten/src/ATen/native/mps/operations/Convolution.mm#L168-L173

is checking all output dims as opposed to just `out_channels`. Assuming the size limitation is only relevant for `out_channels`, we should consider relaxing this guard to only checking it.

### Versions

PyTorch version: 2.6.0a0+gite1196df
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.1.1 (arm64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.0.26.4)
CMake version: version 3.31.1
Libc version: N/A

Python version: 3.12.8 | packaged by conda-forge | (main, Dec  5 2024, 14:19:53) [Clang 18.1.8 ] (64-bit runtime)
Python platform: macOS-15.1.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Max

Versions of relevant libraries:
[pip3] numpy==2.1.3
[pip3] optree==0.13.1
[pip3] torch==2.6.0a0+gite1196df
[pip3] torchaudio==2.5.0a0+ba696ea
[pip3] torchvision==0.20.0a0+e9a3213
[conda] numpy                     2.1.3                    pypi_0    pypi
[conda] optree                    0.13.1                   pypi_0    pypi
[conda] torch                     2.6.0a0+gite1196df           dev_0    <develop>
[conda] torchaudio                2.5.0a0+ba696ea           dev_0    <develop>
[conda] torchvision               0.20.0a0+e9a3213           dev_0    <develop>

**Comment by hvaara** [comment]
I'll investigate and propose a fix if feasible.

@pytorchbot "module: error checking" "module: mps"

**Comment by malfet** [comment]
@jhavukainen can you clarify whether restriction only applies to output channels or to all dimentions?

**Comment by hvaara** [comment]
FWIW we can also get signals from ci, as any version <= 15.0 should be able to repro this, not just macOS 15.0.

**Comment by hvaara** [comment]
It looks like the restriction does not apply solely to output channels, but all dims. Furthermore, I think #140726 introduced a regression.

```python
    def test_conv2d_high_dims(self):
        def helper(weight_sizes, x_sizes):
            weight_cpu = torch.randn(*weight_sizes, device="cpu")
            weight_mps = weight_cpu.detach().clone().to("mps")

            x_cpu = torch.ones(x_sizes, device="cpu")
            x_mps = x_cpu.detach().clone().to("mps")

            y_cpu = F.conv2d(x_cpu, weight_cpu)
            y_mps = F.conv2d(x_mps, weight_mps)
            print(y_cpu)
            print(y_mps)

            self.assertEqual(y_cpu, y_mps, rtol=2.6e-05, atol=2e-04)
        # helper((10, 10, 1, 65535), [1, 10, 1, 65535])
        helper((10, 10, 1, 65536), [1, 10, 1, 65536])
        # helper((100, 100, 1, 65535), [1, 100, 1, 65535])
        # helper((10, 10, 65530, 2), [1, 10, 65530, 2])
        # helper((10, 65536, 10, 10), [1, 65536, 10, 10])
```

Outputs

```
tensor([[[[ 1022.5854]],

         [[  631.4047]],

         [[ -122.7467]],

         [[ 1024.8324]],

         [[ -739.5587]],

         [[   -5.7392]],

         [[  536.1094]],

         [[-1104.8417]],

         [[  349.8428]],

         [[  292.4343]]]])
tensor([[[[0.]],

         [[0.]],

         [[0.]],

         [[0.]],

         [[0.]],

         [[0.]],

         [[0.]],

         [[0.]],

         [[0.]],

         [[0.]]]], device='mps:0')
```

It looks like `nn.Conv1d` is safe with sizes > 2**16, but I need to test this more rigurously. Based on my current tests, `nn.Conv2d` is definitely not -- even on macOS >= 15.1. `nn.Conv3d` needs to be checked as well.

I think we should reintroduce the guard on all macOS versions for the ops that have regressed with high priority before v2.6.0 is shipped. And add more test cases with high dim sizes for the ops that share the conv implementation. This could prevent a regression in the future. I'll create a new issue to track this. Sorry for the breakage.

For this issue, I think the current guard that checks all dim sizes is WAI. But perhaps the error message is a little confusing and should be updated to reflect that it is all dims, not just `out_channels`, that needs to be <= 2**16?
